src/main/scala/zio/keeper/Cluster.scala

```diff
@@ -1,31 +1,267 @@
 package zio.keeper
 
-import java.net.InetAddress
+import java.math.BigInteger
+import java.util.UUID
 
-import zio.{ IO, ZIO }
+import zio._
+import zio.console.putStrLn
+import zio.keeper.Cluster.InternalProtocol.{ Iam, IamNewHere, PleaseIntroduceYourself }
+import zio.keeper.Error.{ HandshakeError, SendError }
+import zio.keeper.Error.NodeUnknown
+import zio.nio._
+import zio.nio.channels.{ AsynchronousServerSocketChannel, AsynchronousSocketChannel }
 import zio.stream.Stream
-import zio.keeper.Cluster._
 
-trait Cluster {
-  final def join: ZIO[Credentials with Discovery with Transport, Error, Unit] = ???
+import scala.reflect.ClassTag
+
+trait Cluster[A] {
+  def nodes: UIO[List[NodeId]]
+
+  def send[B <: A](data: B, receipt: NodeId): IO[Error, Unit]
```

Comment:
We should document the semantics of `send` under conditions of unreachability, downtime, etc. I think it makes sense to be conservative because a user can `.retry`, `.timeout`, etc., already using ZIO, so bundling timeouts and retry logic into `send` don't make sense unless there is a protocol-level reason for doing so.

Reply:
OTOH maybe this is more convenient from user-perspective, especially since the project will have minimal dependencies...

Reply:
I agree, user will be responsible for handles errors and based on their logic retry or set timeout.